### PR TITLE
Add Tracking Networking

### PR DESCRIPTION
**Pull request by ReadeP**
## Overview

Add the ability to track courses and update the UI accordingly.

## Changes Made

- Added tracking ability and UI updates to the HomeViewController
- Fixed a few constraint bugs

## Test Coverage

Manual testing

## Next Steps (delete if not applicable)

Still need to update the UI in the SearchTableViewController when a user tracks a class. Currently the tracking section view is not showing unless the user types in the search bar again.

### CODE REVIEW

File: CourseGrab/ViewControllers/Home/HomeTableViewController.swift [hunk 1]
```diff
@@ -82,29 +86,110 @@ extension HomeTableViewController {
...
+    /// Gets the awaiting sections currently stored locally in the model
+    private func awaitingSections() -> [Section] {
+        if let index = index(of: .awaiting([])),
+            case .awaiting(let sections) = tableSections[index] {
+            return sections
+        }
+        return []
+    }
+
+    /// Gets the inex of a table section or `nil` if it is not in the local model.
```
Comment (danielVebman) [inline CourseGrab/ViewControllers/Home/HomeTableViewController.swift:168]: Typo: `Gets the index...`

File: CourseGrab/ViewControllers/Home/HomeTableViewController.swift [hunk 2]
```diff
@@ -82,29 +86,110 @@ extension HomeTableViewController {
...
+
+    /// Updates the table to reflect the given state.
+    private func show(state: State) {
+        switch state {
+        case .normal:
+            tableView.backgroundView = nil
+        case .empty:
+            tableView.backgroundView = HomeStateView(title: "No Courses Currently Tracked", subtitle: "Tap the search icon to start adding courses", icon: Status.open.icon)
+        case .loading:
+            tableView.backgroundView = HomeStateView(title: "Loading", subtitle: "Fetching your courses", icon: UIImage()) // change icon
```
Comment (danielVebman) [inline CourseGrab/ViewControllers/Home/HomeTableViewController.swift:189]: Delete the comment?

File: CourseGrab/ViewControllers/Home/HomeTableViewHeader.swift [hunk 3]
```diff
@@ -20,13 +20,19 @@ class HomeTableViewHeader: UITableViewHeaderFooterView {
...
+            make.top.equalToSuperview().inset(6)
+        }
     }
 
     required init?(coder: NSCoder) {
         fatalError("init(coder:) has not been implemented")
     }
 
-    func configure(tableSection: HomeTableViewController.TableSection, section: Int) {
+    func configure(tableSection: HomeTableViewController.TableSection) {
```
Comment (danielVebman) [inline CourseGrab/ViewControllers/Home/HomeTableViewHeader.swift:35]: Change to `configure(for tableSection: ...)` to be more consistent with the other configure functions.

File: CourseGrab/Models/Section.swift [hunk 4]
```diff
@@ -56,3 +57,20 @@ struct Section: Codable {
...
+            return data
+        }
+        return Data()
+    }
+
+    func isContentEqual(to source: Section) -> Bool {
+        return differenceIdentifier == source.differenceIdentifier
+    }
+
+}
```
Comment (kevinchan159) [inline CourseGrab/Models/Section.swift:76]: wow neat!

File: CourseGrab/ViewControllers/Home/HomeTableViewController.swift [hunk 5]
```diff
@@ -82,29 +86,110 @@ extension HomeTableViewController {
...
-                    self.tableSections.append(.available(available))
+                // Section identifiers
+                let availableModel = "available"
+                let awaitingModel = "awaiting"
+
+                // Construct source array
+                let currentAvailable = self.availableSections()
+                let currentAwaiting = self.awaitingSections()
+                var source: [ArraySection<String, Section>] = []
+                if currentAvailable.count > 0 {
```
Comment (kevinchan159) [inline CourseGrab/ViewControllers/Home/HomeTableViewController.swift:97]: pref to do `!currentAvailable.isEmpty`

File: CourseGrab/ViewControllers/Home/HomeTableViewController.swift [hunk 6]
```diff
@@ -82,29 +86,110 @@ extension HomeTableViewController {
...
+                    source.append(ArraySection(model: availableModel, elements: currentAvailable))
+                }
+                if currentAwaiting.count > 0 {
```
Comment (kevinchan159) [inline CourseGrab/ViewControllers/Home/HomeTableViewController.swift:100]: same here

File: CourseGrab/ViewControllers/Home/HomeTableViewController.swift [hunk 7]
```diff
@@ -82,29 +86,110 @@ extension HomeTableViewController {
...
+                if available.count > 0 {
+                    target.append(ArraySection(model: availableModel, elements: available))
                 }
-                var newSections: IndexSet = []
-                (0..<self.tableSections.count).forEach { newSections.insert($0) }
-                if newSections.count > 0 {
-                    self.tableView.insertSections(newSections, with: .automatic)
-                } else {
-                    self.showEmptyState()
+                if awaiting.count > 0 {
```
Comment (kevinchan159) [inline CourseGrab/ViewControllers/Home/HomeTableViewController.swift:111]: same here

File: CourseGrab/ViewControllers/Home/HomeTableViewController.swift [hunk 8]
```diff
@@ -127,6 +212,107 @@ extension HomeTableViewController: SPPermissionsDataSource {
...
+    }
+
+    /// Removes a section from the model and returns the change, if any.
+    private func removeSectionFromModel(_ section: Section) -> UpdateDataChange {
+        for (i, tableSection) in tableSections.enumerated() {
+            switch tableSection {
+            case .available(var sections):
+                if let j = sections.firstIndex(where: { $0.catalogNum == section.catalogNum }) {
+                    sections.remove(at: j)
+                    if sections.count > 0 {
```
Comment (kevinchan159) [inline CourseGrab/ViewControllers/Home/HomeTableViewController.swift:289]: use `!isEmpty`

File: CourseGrab/ViewControllers/Search/SearchTableViewController.swift [hunk 9]
```diff
@@ -59,9 +60,11 @@ extension SearchTableViewController {
     private func textDidChange(_ textField: UITextField) {
         // TODO: Ignore old requests and skip every 0.2 seconds
         guard let text = textField.text, text.count > 2 else {
-            DispatchQueue.main.async {
-                self.courses.removeAll()
-                self.tableView.reloadData()
+            if courses.count > 0 {
```
Comment (kevinchan159) [inline CourseGrab/ViewControllers/Search/SearchTableViewController.swift:63]: same here

File: CourseGrab/ViewControllers/Search/SearchTableViewController.swift [hunk 10]
```diff
@@ -128,6 +172,7 @@ extension SearchTableViewController {
     
     override func viewDidAppear(_ animated: Bool) {
         textField.becomeFirstResponder()
+        textDidChange(textField)
```
Comment (kevinchan159) [inline CourseGrab/ViewControllers/Search/SearchTableViewController.swift:175]: whats this for?

**Pull request merged**